driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/BatchCursorPublisher.java

```diff
@@ -0,0 +1,169 @@
+import com.mongodb.internal.async.AsyncBatchCursor;
+import com.mongodb.internal.operation.AsyncReadOperation;
+import com.mongodb.internal.operation.Operations;
+import com.mongodb.lang.Nullable;
+import com.mongodb.reactivestreams.client.ClientSession;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+import static com.mongodb.assertions.Assertions.notNull;
+
+abstract class BatchCursorPublisher<T> implements Publisher<T> {
+    private final ClientSession clientSession;
+    private final MongoOperationPublisher<T> mongoOperationPublisher;
+
+    private Integer batchSize;
+
+    BatchCursorPublisher(@Nullable final ClientSession clientSession, final MongoOperationPublisher<T> mongoOperationPublisher) {
+        this.clientSession = clientSession;
+        this.mongoOperationPublisher = notNull("mongoOperationPublisher", mongoOperationPublisher);
+    }
+
+    abstract AsyncReadOperation<AsyncBatchCursor<T>> asAsyncReadOperation();
+
+    AsyncReadOperation<AsyncBatchCursor<T>> asAsyncFirstReadOperation() {
+        return asAsyncReadOperation();
+    }
+
+    @Nullable
+    ClientSession getClientSession() {
+        return clientSession;
+    }
+
+    MongoOperationPublisher<T> getMongoOperationPublisher() {
+        return mongoOperationPublisher;
+    }
+
+    Operations<T> getOperations() {
+        return mongoOperationPublisher.getOperations();
+    }
+
+    MongoNamespace getNamespace() {
+        return mongoOperationPublisher.getNamespace();
+    }
+
+    ReadPreference getReadPreference() {
+        return mongoOperationPublisher.getReadPreference();
+    }
+
+    CodecRegistry getCodecRegistry() {
+        return mongoOperationPublisher.getCodecRegistry();
+    }
+
+    boolean getRetryReads() {
+        return mongoOperationPublisher.getRetryReads();
+    }
+
+    Class<T> getDocumentClass() {
+        return mongoOperationPublisher.getDocumentClass();
+    }
+
+
+    @Nullable
+    public Integer getBatchSize() {
+        return batchSize;
+    }
+
+    public Publisher<T> batchSize(final int batchSize) {
+        this.batchSize = batchSize;
+        return this;
+    }
+
+    public Publisher<T> first() {
+        return batchCursor(this::asAsyncFirstReadOperation)
+                .flatMap(batchCursor -> Mono.create(sink -> {
+                    batchCursor.setBatchSize(1);
+                    Mono.from(batchCursor.next())
+                            .doOnTerminate(batchCursor::close)
+                            .doOnError(sink::error)
+                            .doOnSuccess(results -> {
+                                if (results == null || results.isEmpty()) {
+                                    sink.success();
```

Comment:
Let's unit test this branch.  It's the only uncovered code in this file

Reply:
Done